Diana

koshreality opened the pull request.

koshreality committed:
скрипт для подсчета баланса

koshreality committed:
Тест баланса

Сценарий: банк переводит поставщику 100 у.е. с категорией "Аванс на материалы".
В результате поставщик получает на свой баланс 100 у.е., т.к. на категорию "Аванс на материалы" не установлен флаг на неучастие в прибыли (не равен 1), и, соответственно, данная категория будет участвовать в прибыли - это и есть достаточное условие для увеличения переменной Profit; поставщик указан в качестве Payee и не входит в список работников, что является достаточным условием для неизменения переменных Cost  и EmployeeCost.
Банк теряет 100 у.е. с баланса, т.к. указан в качестве Payer и не входит в список работников (что является достаточным условием для изменения Cost и неизменения EmployeeCost).

sevryukov commented:
@koshreality Сценарий описан подробно и верно, спасибо.
Как выглядит сам процесс тестирования? Что и в какой последовательности делаете, как приходите к выводу о том, что тест прошёл или нет?

sevryukov commented on Tokarev_Kayumova/balance.sql:
```diff
@@ -0,0 +1,21 @@
+-- Payee.Oid in (Select Oid from Employee)
+
+
+SELECT NotInPaymentParticipantProfit from PaymentCategory where Name = '����� �� ���������';
```
@koshreality Почему возникают проблемы с кодировками не разбирались?

koshreality replied:
С кодировкой не разбиралась, наверное, проблемы возникают из-за несоответствия кодировки, которую использует гитхаб, и той, в которой сохраняются sql-файлы в SSMS 
(я не нашла настроек в студии, позволяющих изменить ее)

koshreality commented:
@sevryukov 
Последовательность тестирования проходит следующим образом: сначала исполняется balance.sql для вывода в консоль информации о балансе и свойствах участников, влияющих на результат платежа, 
затем выполняются insert_payment.sql и снова balanse.sql для повторной проверки баланса
Проверяла корректность теста вручную, сравнивала фактическое изменение баланса с ожидаемым

sevryukov commented:
@koshreality Есть идеи, как можно упростить процесс? Представьте, что тестов будет не один, а значительно больше? Или представьте, что тест нужно выполнить несколько раз (например, до и после изменений в функциях)?
Я не прошу переделать, а прошу дать оценку.

koshreality commented:
Можно оформить в виде хранимой функции, принимающей на вход параметры insert'а, то есть Amount, Category, Payer, Payee, а так же ожидаемое изменение баланса участников платежа, и сначала, перед выполнением insert'a, сохранять в переменных баланс, затем выполнять insert, снова запросить баланс, сравнить его изменение с ожидаемым и выдавать true или false

sevryukov commented:
@koshreality Начало хорошее, но можно было бы улучшить. Либо подумайте и напишите как, либо обсудим при встрече.
Спасибо.

koshreality commented:
Можете привести какой-нибудь тестовый сценарий, чтобы я поняла, что нужно улучшать?

sevryukov commented:
@koshreality Сценарий может быть любым. Вопрос не в контексте сценария, а в контексте процесса тестирования.

koshreality commented:
@sevryukov Если честно, я не очень понимаю, в чем именно требуются улучшения. Мне кажется, что если есть список тестов "данные-результат", то, запустив в цикле такую функцию для каждого теста и как-то обработав результаты (например, посчитав количество успехов), мы и получим требуемый результат

sevryukov commented:
Ок, обсудим при встрече.

ср, 2 мая 2018, 19:48 koshreality <notifications@github.com>:

> @sevryukov <https://github.com/sevryukov> Если честно, я не очень
> понимаю, в чем именно требуются улучшения. Мне кажется, что если есть
> список тестов "данные-результат", то, запустив в цикле такую функцию для
> каждого теста и как-то обработав результаты (например, посчитав количество
> успехов), мы и получим требуемый результат
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/sevryukov/DBOpt_Course_PaymentData/pull/8#issuecomment-386044624>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAsa73DQhAwk5kEB1QbZYQRfyjNF5TPiks5tueNkgaJpZM4Tq1Xb>
> .
>

koshreality committed:
Task 3

Сценарий отложенного вычисления:
т.к. в триггере производится полный пересчет баланса, то, если вставить, например, 5 тысяч записей с новым платежом без триггера, а потом одну с ним, то результат будет таким же, как если бы считать баланс после каждой вставки, в связи с чем сценарий отложенного вычисления заключается в отключении этого триггера, включении его после вставки записей, кроме одной, и затем вставки этой записи с включенным триггером.
в скрипте triggeronoff.sql осуществляется проверка предположения об ускорении вставки при отключении триггера.
Для этого производится повторная вставка последних 5000 записей без триггера и с триггером. Время работы вставки записей с триггером — 40 секунд, а без — меньше секунды.
в PaymentData.java: вставка 5000 записей (я использовала ту же, что я сдавала в прошлый раз) по одной, триггер работает после каждой вставки — 336 секунд;
вставка 5000 записей батчем — 876 секунд (непонятно, почему);
вставка 4999 записей по одной с выключенным триггером, а затем одной с включенным для пересчета баланса — 4.3 секунды;
вставка 4999 записей батчем с выключенным триггером, а затем одной с включенным для пересчета баланса — 3 секунды.

koshreality commented:
Сценарий отложенного вычисления: 
т.к. в триггере производится полный пересчет баланса, то, если вставить, например, 5 тысяч записей с новым платежом без триггера, а потом одну с ним, то результат будет таким же, как если бы считать баланс после каждой вставки, в связи с чем сценарий отложенного вычисления заключается в отключении этого триггера, включении его после вставки записей, кроме одной, и затем вставки этой записи с включенным триггером.
в скрипте triggeronoff.sql осуществляется проверка предположения об ускорении вставки при отключении триггера. 
Для этого производится повторная вставка последних 5000 записей без триггера и с триггером. Время работы вставки записей с триггером — 40 секунд, а без — меньше секунды. 
в PaymentData.java: вставка 1000 записей (я использовала ту же, что я сдавала в прошлый раз) по одной, триггер работает после каждой вставки — 336 секунд; 
вставка 1000 записей батчем — 876 секунд (непонятно, почему);
вставка 999 записей по одной с выключенным триггером, а затем одной с включенным для пересчета баланса — 4.3 секунды; 
вставка 999 записей батчем с выключенным триггером, а затем одной с включенным для пересчета баланса — 3 секунды.

sevryukov commented:
@koshreality Интересные эксперименты и наблюдения. Для полноты картины добавьте пожалуйста выводы (можно отредактировать Ваш комментарий или добавить новый).

NIKKAS29 commented:
Прошу прощения, что влезаю в этот PullRequest. Просто была вот прям идентичная мысль об оптимизации, но не проверялась на практике, т.к. триггер T_Payment_AI обновляет же балансы в строках таблицах-участницах конкретного платежа? Таким образом все вставки платежей с отключенным этим триггером просто будут не учтены. Или все таки будут?
Если мой комментарий не уместен в данном обсуждении, дайте знать - удалю

sevryukov commented:
@NIKKAS29 Не извиняйтесь, я только приветствую такое взаимодействие.
@koshreality Буду признателен, если поддержите диалог с коллегой.

koshreality committed:
Test 3 upd

1 by 1 with trigger: 5098
Batch with trigger: 5049
1 by 1 without trigger: 2504
Batch without trigger: 2494
время работы в милисекундах для двух записей

koshreality commented:
@NIKKAS29 При вставке сразу пачки записей можно вызывать функцию пересчета баланса (F_CalculatePaymentParticipantBalance) для каждого участника один раз, даже если это позволит не выполнять триггер хотя бы один раз, все равно будет выигрыш во времени, что я проверила и обновила код. В нем выполняется сравнение как раз для такого случая. Выигрыш составил 2.5 секунды.

@sevryukov Выводы:
1) при выполнении предложенного сценария отложенных вычислений получаем существенный выигрыш во времени. В случае двух вставок записей с выполнением триггера один раз, то этот выигрыш составляет около 50%.
2) К недостаткам стоит отнести то обстоятельство, что надо знать id всех  участников всех платежей и однократно запускать функцию пересчета баланса для каждого из них после вставки всех записей в рамках одной и той же транзакции

sevryukov commented:
@koshreality Спасибо за уточнения
